File: sdk/typescript/packages/api/src/context.ts
export interface AnalyticsContext {
  userId?: string;
  anonymousId?: string;
  attributes?: Record<string, unknown>;
}

// merge combines base and override AnalyticsContexts.
// Non-empty override fields win. Attributes merged key-by-key with override keys winning.
export function merge(base: AnalyticsContext, override: AnalyticsContext): AnalyticsContext {
  const result: AnalyticsContext = {
    userId: override.userId || base.userId,
    anonymousId: override.anonymousId || base.anonymousId,
  };

  if (base.attributes || override.attributes) {
    result.attributes = { ...base.attributes, ...override.attributes };
  }

  return result;
}

File: sdk/kotlin/api/src/main/kotlin/io/eventspec/analytics/Context.kt
package io.eventspec.analytics

data class AnalyticsContext(
    val userId: String = "",
    val anonymousId: String = "",
    val attributes: Map<String, Any?> = emptyMap(),
)

// merge combines base and override. Non-empty override fields win.
// Attributes are merged key-by-key with override keys winning.
fun merge(base: AnalyticsContext, override: AnalyticsContext): AnalyticsContext {
  val userId = override.userId.ifEmpty { base.userId }
  val anonymousId = override.anonymousId.ifEmpty { base.anonymousId }
  val attributes = base.attributes + override.attributes
  return AnalyticsContext(userId, anonymousId, attributes)
}
